Move shared component mixins to global_styles

johnbarrierwilson opened the pull request:
### Summary

In order to create themes more easily, this PR will move component mixins that are used in more than one component—for example, euiPanel is used in the panel, card and droppable components—to the `src/global_styling/mixins` directory. Practically, this sets the foundation for doing more extensive work for EUI Amsterdam.

### Mixins

Mixins that need moving to `src/global_styling`:
- [x] `components/button/_mixins.scss`
- [x] `components/form/_mixins.scss`
    - [x] `components/form/form_control_layout/_mixins.scss`
    - [x] `components/form/switch/_mixins.scss`
- [x] `components/loading/_mixins.scss`
- [x] `components/panel/_mixins.scss`
- [x] `components/popover/_mixins.scss`

### Checklist

- [x] Checked in **dark mode**
- [x] Checked in **mobile**
- [x] Checked in **IE11** and **Firefox**
~- [ ] Props have proper **autodocs**~
~- [ ] Added **documentation** examples~
~- [ ] Added or updated **jest tests**~
- [x] Checked for **breaking changes** and labeled appropriately
~- [ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately

johnbarrierwilson commented:
A quick note: once I got into the `form` mixin stuff, it adversely required the variables and mixins of many more form-based components. This included: form, form_control_layout, range, and switch mixins and variables.

johnbarrierwilson commented:
@cchaos @ryankeairns (and others) — As previously discussed, I've finished up pulling out any _shared_ mixins from the components and put them into the `/global_styles` directory. I had a few thoughts at this point, so instead of doing it through comments, here's a quick screencast:

https://www.loom.com/share/8095521ed8e940fea659ae35442aa5ed

Link to the Whimsical map: https://whimsical.com/NgtG9QtjSxn7qYP2b3yKEw

#### Questions
1. Is it worth going down this route?
2. If so, should we do it in this PR or merge and start a new one?
3. If not, just merge and continue on as-is?

ryankeairns commented:
@johnbarrierwilson I appreciate you digging into and grokking the world view of the `/src` directory, this map is helpful, thank you! 😄 

My initial reaction, regardless of whether we tackle this re-organization or not, is that it feels wise to merge this PR with its current smaller scope - keeping it small and digestible while allowing you to keep momentum on the visual design effort.

Personally, I like the convenience of having the SCSS files right there with the component from a development perspective as it's very easy to find files/styles.

This proposed change also has me wondering about potential impacts this might have to our current build/release process (maybe it's minimal?) and what (if any) affect this could have on our future plans with regards to modularization/use of astroturf, etc. In other words, would moving these files have any impact down the road? Tagging in Chandler and Greg for more perspective on those fronts.

cc:/ @thompsongl @chandlerprall

johnbarrierwilson commented:
@ryankeairns Glad the map is helpful! 😊

> ...it feels wise to merge this PR with its current smaller scope...

I totally agree and will do some more testing to make sure everything is stable, then I'll open it for review.

> ...has me wondering about potential impacts...

I think those are also valid concerns. The largest issue I was attempting to address here is the ease of theme-ability by trying to avoid this as illustrated in my previous PR:

![image](https://user-images.githubusercontent.com/2472911/69594912-6e318580-0fc3-11ea-818c-f3ade6b3c512.png)

From personal experience, I've used a `/styles` directory for many larger projects at this size, and I found it much easier to balance the scale using this "separate concerns" philosophy. That said, I'm still not as familiar with EUI as everyone else here, so I'll make sure to head any advice/concerns this may raise.

cchaos commented:
@johnbarrierwilson I agree with @ryankeairns about having the component SASS files next to the component JS files. We did this very intentionally so that we knew exactly where to look for styles and we've matched our SASS file namings based on the JS files. We also carry this pattern to Kibana as that file structure is way less straight forward and plugins having their own compile file.

It does cause some weirdness with globals and shared styles/mixins. However, the tradeoff is not having to scour the files or searching in project for classes to find the right files. In all my previous projects I have worked the way you mentioned (styles in one folder, js in another) and I'm a complete convert to this pattern as a maintainer of such a large system and working within and gigantic file system like Kibana.

We are definitely happy to help you reorganize certain mixins to globals, but we will need to keep the general structure as it is.

johnbarrierwilson commented:
👍🏻10-4! Thanks for shedding a bit more light on the thoughts. I'll move forward with this PR as-is!

thompsongl commented:
Whoa, I really appreciate all your thought and explanation here.

The gist is, I agree that component-specific stylesheets should remain inside the same directory as the JS/TS component files, and we should reorganize some shared/global things.

Potential impacts to the _current_ build system aren't concerning (even in the proposed reorg), but impacts to an eventual move to CSS Modules or CSS-in-JS would be fairly substantial. The main reason being that with either new approach, there will be actual interop between style files and component files (i.e., the component file will directly import the style file instead of a global manifest).

The argument for separating concerns requires more discussion. Separating _technologies_ (HTML, CSS, JS, etc.,) is likely the wrong place for us to draw boundaries. In EUI, component state, logic, and appearance are fundamentally intertwined and require multiple technologies to accomplish the final rendered output. EUI concerns are more appropriately separated at the _component_ level. The way we do Sass at the moment obscures this a bit, but it will become more clear as modularity is introduced.

That being said, globals are tricky in any case for EUI, and I think the current state of this PR helps address some of those and raises some really good discussion as we move toward future system choices.

johnbarrierwilson commented:
That's a fantastic summary @thompsongl! Thanks for taking the time to pitch in!

cchaos commented:
@johnbarrierwilson Is this ready for review? Just wasn't sure based on your comment above:

> will do some more testing to make sure everything is stable, then I'll open it for review.

johnbarrierwilson commented:
@cchaos Sorry, was out on Wednesday. Yes, this is ready for review now.

cchaos commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 - Added `disabled` prop to the `EuiCheckboxGroup` definition ([#2545](https://github.com/elastic/eui/pull/2545))
 - Added `disabled` option to the `option` attribute of the `options` object that is passed to the `EuiCheckboxGroup` so that checkboxes in a group can be individually disabled ([#2548](https://github.com/elastic/eui/pull/2548))
 - Added `EuiAspectRatio` component that allows for responsively resizing embeds ([#2535](https://github.com/elastic/eui/pull/2535))
+- Moved any shared component-level Sass variables and mixins into the `global_styling` directory ([#2551](https://github.com/elastic/eui/pull/2551))
```
You'll want to add this under a new heading of `**Breaking changes**`

cchaos commented on src/components/button/_index.scss:
```diff
@@ -1,5 +1,4 @@
 @import 'variables';
```
There are variables in here that are referenced in the buttons mixin file that need to be moved to global.

cchaos commented on src/components/facet/_index.scss:
```diff
@@ -1,5 +1,4 @@
 @import '../button/variables';
```
Yeah, since this variables file is referenced in multiple files it should be moved to globals.

cchaos commented on src/global_styling/variables/_index.scss:
```diff
@@ -14,3 +14,7 @@
 @import 'z_index';
+
+@import 'form';
+@import 'form_control_layout';
```
These two can probably be combined into one file since it's probably confusing as to what the difference is.

cchaos commented on src/components/card/_card.scss:
```diff
@@ -1,4 +1,3 @@
-@import '../panel/mixins';
 @import '../badge/beta_badge/mixins';
```
Another contender for being moved to global?

cchaos commented on src/global_styling/mixins/_form.scss:
```diff
@@ -1,6 +1,6 @@
-@import 'form_control_layout/mixins';
+@import '../variables/form';
+@import '../variables/form_control_layout';
+@import 'form_control_layout';
```
It shouldn't be necessary to re-import the variables here as they're part of the global manifest. It also then might be safe to also combine the `form_control_layout` mixins into one file.

johnbarrierwilson commented:
Alrighty, those changes should be good-to-go. 
- Moved the button mixins and variables
- Consolidated the form variables files
- Moved beta_badge mixins
- Moved tooltip variables and mixins
- Refactored panel mixin to require entire selector (`.` + `string`)
- Removed $euiPanelBorderColor

johnbarrierwilson commented:
@cchaos Awesome! Thanks for walking me through my first contribution to EUI! 😊
